design/pipeline-caching.md

```diff
@@ -0,0 +1,157 @@
+# Pipeline Caching
+
+*** Status: Work in Progress ***
+
+We have observed the need for Azure Pipelines to implement a dependency caching mechanism that allows the outputs of steps within a pipeline to be optimized (or even skipped) if a suitable cached version of the outputs of those steps already exists. Caching can be effective provided the cost of determining a cache hit and acquiring the contents of the cache is cheaper than the cost of producing the output again from scratch.
+
+## Proposal
+
+This proposal contains a number of interconnected elements. We have attempted to layer them in this document from the most primitive building blocks up to simplified YAML syntax to make caching "just work" in the majority of cases. Our approach to building a caching mechanism for Azure Pipelines will be to get the fundamental building blocks right and prove we can increase build performance and then evolve the YAML syntax once we know exactly what that syntax will need to describe.
+
+### Restore and Save Cache Tasks
+
+The basic functions of Pipeline Cache will be implemented by two tasks. The ```RestoreCache``` task and the ```SaveCache``` task. Both tasks will specify a collection of files and environment variables which can be used as a key to lookup the cache and identify a hit or a miss. Each task will also take a collection of paths which will be fetched from or stored in the cache associated with that key.
+
+Consider the following Azure Pipelines pipeline defined in YAML:
+
+```yaml
+steps:
+- script: yarn
+- script: yarn build
+```
+
+In this example we are working on a React-based application which was bootstrapped using ```create-react-app```. The output of this command is a directory with about ~28,000 files including the ```node_modules``` directory. On a local developer machine the ```yarn``` command will trigger the installation of any missing packages. Typically this will execute quickly because all packages already exist on the developer machine. However in cloud-based CI environments where build machines are recycled after every build there is no persistent state which means that Yarn must download all packages from scratch.
+
+Applying caching to this pipeline would involve adding a restore cache step and save cache step before and after the ```-script: yarn``` step.
+
+```yaml
+steps:
+- task: RestoreCache@0
+  inputs:
+    keys: |
+      package.json
```

Comment:
How are we going to determine this key is a file to hash vs a string value to hash?

Reply:
Initially we had thought about having some kind of expression syntax here, something like ```$(filehash:package.json)```. Given we don't support expressions like this at the moment I thought t might be simpler just try to detect whether it is a file or an environment variable.

We might look for a file first, and then check for environment variables - does that seem reasonable?

I'd like to avoid inventing an expression syntax just for Pipeline Caching scenarios. It sounded a bit challenging to lift what was done for task conditionals, especially given those expressions only return Booleans.

Reply:
So concretely, the resolution would be:

1. Check if it is a file or directory and then hash
2. Check if it is an environment variable then hash.
3. Treat it as a string and hash it/add it to the key.

Reply:
My preference is to err on the side of caution here so that a typo'd filename would result in an error instead of false cache hits.

Reply:
We could take another look at putting an expression parser agent-side. I agree it's kind of a big lift for something that should be simpler.

Is it enough to add a sigil at the front of each line and parse with a regex:

```yaml
keys: |
  file:package-lock.json
  env:MY_ENV_VAR
  str:salty-key
```

with optional handling of non-sigil'd lines - maybe assume they're strings?

Reply:
How are the keys processed here?  Are they tried in order with and the instructions for the user are to put the most specific first and we will fall back through the rest?

Reply:
Seems a bit too much like magic numbers. What about just keep it easy with 3 separate properties. More can be added as needed rather than supporting more custom expression parsers, e.g.:

```yaml
keys:
  files: package-lock.json # or could be string literal, as well as everything below
  env: MY_ENV_VAR
  strings: salty-key
```
This would also work better with your azure-pipelines syntax highlighter for VSCode (though a YAML schema would be so much nicer, which seems some people are using JSON schema as an implementation).

Reply:
+1 on `key` to disambiguate if we later want to support multiple keys

Reply:
OK - so the way I think this will work is that we'll have a single property called ```key:``` and then by default we assume that anything listed is a file (the most common case). If the desire is to set a key using an environment variable then you can prefix the line with ```env:``` to signal that it should be processed as an environment variable.

String keys are pretty much pointless so I think I might just ignore that scenario for now, but if I was to do it we could either support a ```string:``` prefix or interpret something like ```'somestring'``` as a string literal - not sure if that is worth it or not though. I'll update the PR and we can do another round of comments on the result.

My reason for going with the prefix option is that if we don't the fully fleshed out task syntax is pretty ugly (we would need multiple input properties for each key type). Doesn't seem worth it given files will be the most common scenarios.

Keep in mind that we will add some additional info the key such as OS version (thoughts?).